Get version string from git-describe, git-archive or CMake config

jschmidt-icinga opened the pull request:
Currently if the source directory is not a git repo, the version string will be the last released version set in the `ICINGA2_VERSION` file.

This PR makes the following improvements:
+ If `git-describe` fails because the source directory isn't a git repo or `ICINGA2_GIT_VERSION_INFO` is OFF, the version is read from a file processed by `git-archive`.
+ The version file `ICINGA2_VERSION` is removed and replaced with a variable at the top of the main CMakeLists.txt that will need to be updated manually on a new release.
+ New variables `ICINGA2_PACKAGE_(VERSION|REVISION|VENDOR)` are added that can be customized by package build process. These are utilized for Windows MSI builds (via CPACK)
+ Package version information can be printed in `icinga2 --version` if the new option `ICINGA2_INCLUDE_PACKAGE_INFO` is set to ON (defaults to OFF). I added an additional option for this because I wanted to use defaults, description and type for the package version/revision/vendor cache variables, and otherwise there is no reliable way to detect that the user has changed the information and include the `--version` section based on that.

Here is an overview of how the variables/constants and their meaning has changed:

Before:

| CMake | C++ (and Windows cruft) | meaning |
| - | - | - |
| GIT_VERSION | VERSION | The best approximation of the actual version, either git describe or spec version + revision from the ICINGA2_VERSION file |
| ICINGA2_VERSION | ICINGA2_VERSION | Like above with the 'v' prefix removed and no package revision suffix in case no git version was found |
| SPEC_VERSION | SPEC_VERSION | The package version defined manually in the ICINGA2_VERSION file  |

After:

| CMake | C++ (and Windows cruft) | meaning |
| - | - | - |
| ICINGA2_VERSION | ICINGA_VERSION | The best approximation of the actual version, going from git-describe -> git-archive info -> package version + revision |
| ICINGA2_RC_VERSION | ICINGA_RC_VERSION | Like above, but with removed 'v' prefix |
| ICINGA2_PACKAGE_VERSION | ICINGA_PACKAGE_VERSION | The manually defined, abbreviated version string |


Fixes #10016
Fixes #8960
Closes #9649

julianbrost commented on icinga-version.h.in:
I don't see why this shouldn't just become part of https://github.com/Icinga/icinga2/blob/master/config.h.cmake

That's already used for passing all kinds of information from CMake to C++.

julianbrost reviewed (commented):
You found a thing here that I wasn't really aware of so far: `GetAppSpecVersion()`. It looks like you started using that new package version thing where this was used before, can you summarize why you did it that way? That makes the fixes for #10016 and #8960 more interleaved than I had expected: I would have expected that package version to be an informative thing for the `icinga2 --version` output only, so can you please elaborate why you chose to use it in more places?

> Please tell me if a more complex approach is needed.

If anything, this should become simpler :D

jschmidt-icinga committed:
Rename `GetAppSpecVersion()` to `GetAppPackageVersion()`

jschmidt-icinga committed:
Use `Application::GetAppVersion()` consistently

jschmidt-icinga committed:
Generate version information file during git-archive

jschmidt-icinga committed:
Get version string from git-describe, git-archive or CMake config

jschmidt-icinga commented:
+ Rebased the PR, and added an additional comment where something was unclear to me after a few months of not looking at this.
+ Also tried to split it into multiple commits, but most of the changes are interdependent, so I was only able to break off three small changes.
